File: scripts/step2_rule_judge.mjs
import { mkdir, readFile, readdir, writeFile } from "node:fs/promises";
import path from "node:path";

const CHOICE_PATTERN = /\b([A-F])\b/i;

function parseArgs(argv) {
  const args = {
    modelChatFilesFolder: path.resolve("Chat_results"),
    ruleJudgeOutputFolder: path.resolve("Rule_Judge"),
    evalQaRoot: path.resolve("."),
  };
  for (let i = 0; i < argv.length; i += 1) {
    const key = argv[i];
    const value = argv[i + 1];
    if (key === "--model_chat_files_folder") args.modelChatFilesFolder = path.resolve(value);
    if (key === "--rule_judge_output_folder") args.ruleJudgeOutputFolder = path.resolve(value);
    if (key === "--Eval_QA_root") args.evalQaRoot = path.resolve(value);
    if (key.startsWith("--")) i += 1;
  }
  return args;
}

function extractChoice(text) {
  const normalized = String(text ?? "").trim().toUpperCase();
  const match = normalized.match(CHOICE_PATTERN);
  if (match) return match[1].toUpperCase();
  return normalized.slice(0, 1);
}

async function loadQa(evalQaRoot, datasetName) {
  const candidates = [
    path.join(evalQaRoot, "Eval_QA", `${datasetName}_QA_sample.json`),
    path.join(evalQaRoot, "Eval_QA", `${datasetName}_QA.json`),
  ];
  for (const file of candidates) {
    try {
      return JSON.parse(await readFile(file, "utf8"));
    } catch (error) {
      if (error.code !== "ENOENT") throw error;
    }
  }
  throw new Error(`Cannot find QA file for ${datasetName}`);
}

const args = parseArgs(process.argv.slice(2));
await mkdir(args.ruleJudgeOutputFolder, { recursive: true });
const files = (await readdir(args.modelChatFilesFolder))
  .filter((name) => name.endsWith("_eval.json"))
  .map((name) => path.join(args.modelChatFilesFolder, name));

for (const file of files) {
  const datasetName = path.basename(file).replace("_eval.json", "");
  const qaData = await loadQa(args.evalQaRoot, datasetName);
  const evalData = JSON.parse(await readFile(file, "utf8"));
  const outDir = path.join(args.ruleJudgeOutputFolder, path.basename(file).replace("_eval.json", "_rule_eval"));
  await mkdir(outDir, { recursive: true });

  for (const [qidVid, item] of Object.entries(evalData)) {
    const qid = String(item.qid ?? qidVid.split("_")[0]);
    const answer = qaData[qid]?.answer ?? "";
    const prediction = extractChoice(item.output_sequence);
    const judged = {
      ...item,
      output_rule_choice: prediction,
      answer,
      correct: prediction === answer,
    };
    await writeFile(path.join(outDir, `${qidVid}.json`), JSON.stringify({ [qidVid]: judged }, null, 2), "utf8");
  }
  console.log(`${file} is finished.`);
}

File: Step2_rule_judge.py
# -*- coding: utf-8 -*-
"""Rule-based judge for ChronoPhyBench-style multiple-choice outputs.

This follows the same Step2 role as Video-Bench's ChatGPT/T5 judge scripts, but
keeps the default evaluator offline and deterministic.
"""

import argparse
import glob
import json
import os
import re


CHOICE_PATTERN = re.compile(r"\b([A-F])\b", re.IGNORECASE)


def extract_choice(text):
    if text is None:
        return ""
    text = str(text).strip()
    if not text:
        return ""
    match = CHOICE_PATTERN.search(text.upper())
    if match:
        return match.group(1).upper()
    return text[:1].upper()


def load_qa(eval_qa_root, dataset_name):
    qa_path = os.path.join(eval_qa_root, "Eval_QA", f"{dataset_name}_QA_sample.json")
    if not os.path.exists(qa_path):
        qa_path = os.path.join(eval_qa_root, "Eval_QA", f"{dataset_name}_QA.json")
    with open(qa_path, "r", encoding="utf-8") as f:
        return json.load(f)


def process_file(eval_file, args):
    dataset_name = os.path.basename(eval_file).replace("_eval.json", "")
    qa_data = load_qa(args.Eval_QA_root, dataset_name)
    with open(eval_file, "r", encoding="utf-8") as f:
        eval_data = json.load(f)

    output_folder = os.path.join(
        args.rule_judge_output_folder, os.path.basename(eval_file).replace("_eval.json", "_rule_eval")
    )
    os.makedirs(output_folder, exist_ok=True)

    for qid_vid, item in eval_data.items():
        qid = str(item.get("qid", qid_vid.split("_")[0]))
        gt = qa_data[qid].get("answer", "")
        pred = extract_choice(item.get("output_sequence", ""))
        judged = dict(item)
        judged["output_rule_choice"] = pred
        judged["answer"] = gt
        judged["correct"] = pred == gt
        output_file = os.path.join(output_folder, f"{qid_vid}.json")
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump({qid_vid: judged}, f, indent=2, ensure_ascii=False)
    print(f"{eval_file} is finished.")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--model_chat_files_folder", type=str, default="./Chat_results")
    parser.add_argument("--rule_judge_output_folder", type=str, default="./Rule_Judge")
    parser.add_argument("--Eval_QA_root", type=str, default="./")
    args = parser.parse_args()

    os.makedirs(args.rule_judge_output_folder, exist_ok=True)
    evaljson_list = glob.glob(f"{args.model_chat_files_folder}/*_eval.json")
    for eval_file in evaljson_list:
        process_file(eval_file, args)


if __name__ == "__main__":
    main()
